fn/rwguard.go

```diff
@@ -0,0 +1,25 @@
+package fn
+
+import "sync"
+
+type RwGuard[A any] struct {
```

Comment:
Current revision of this PR doesn't make use of this API. I'm leaving it here until this PR is finalized and if it doesn't make the final cut, I can drop the commit entirely or I can PR separately if we think it is useful.